crates/extensions/c8y_mapper_ext/src/operations/mod.rs

```diff
@@ -13,47 +13,270 @@
+    pub capabilities: Capabilities,
+    pub auto_log_upload: AutoLogUpload,
+    pub tedge_http_host: Arc<str>,
+    pub tmp_dir: Arc<Utf8Path>,
+    pub mqtt_schema: MqttSchema,
+    pub c8y_prefix: tedge_config::TopicPrefix,
+    pub software_management_api: SoftwareManagementApiFlag,
+    pub command_id: IdGenerator,
+
+    pub http_proxy: C8YHttpProxy,
+    pub c8y_endpoint: C8yEndPoint,
+    pub auth_proxy: ProxyUrlGenerator,
+
+    pub downloader: ClientMessageBox<IdDownloadRequest, IdDownloadResult>,
+    pub uploader: ClientMessageBox<IdUploadRequest, IdUploadResult>,
+    pub mqtt_publisher: LoggingSender<MqttMessage>,
+
+    pub running_operations: Arc<Mutex<HashMap<Arc<str>, RunningOperation>>>,
+}
+
+impl OperationHandler {
+    /// Handles an MQTT command id message.
+    ///
+    /// All MQTT messages with a topic that contains an operation id, e.g.
+    /// `te/device/child001///cmd/software_list/c8y-2023-09-25T14:34:00` need to be passed here for
+    /// operations to be processed. Messages not related to operations are ignored.
+    ///
+    /// `entity` needs to be the same entity as in `message`.
+    ///
+    /// When a message with a new id is handled, a task will be spawned that processes this
+    /// operation. For the operation to be completed, subsequent messages with the same command id
+    /// need to be handled as well.
+    ///
+    /// When an operation terminates (successfully or unsuccessfully), an MQTT operation clearing
+    /// message will be published automatically, which does not need to be handled.
+    pub async fn handle(self: &Arc<Self>, entity: Entity, message: MqttMessage) {
+        let Ok((_, channel)) = self.mqtt_schema.entity_channel_of(&message.topic) else {
+            return;
+        };
+
+        let Channel::Command { operation, cmd_id } = channel else {
+            return;
+        };
+
+        let message = OperationMessage {
+            operation,
+            entity,
+            cmd_id: cmd_id.into(),
+            message,
+        };
+
+        let topic: Arc<str> = message.message.topic.name.clone().into();
+        let running_operation = {
+            let mut lock = self.running_operations.lock().unwrap();
+            let op = lock.get(&topic);
+
+            if let Some(running_operation) = op {
+                // task already terminated
+                if running_operation.tx.send(message).is_err() {
+                    let running_operation = lock.remove(&topic).unwrap();
+                    Some(running_operation)
+                } else {
+                    None
+                }
+            } else {
+                let handler = self.clone();
+                let running_operation = RunningOperation::spawn(message, handler);
+
+                lock.insert(topic, running_operation);
+                None
+            }
+        };
+
+        if let Some(running_operation) = running_operation {
+            let join_result = running_operation.handle.await;
```

Comment:
This could be my misunderstanding of the `await` contract on a tokio `JoinHandle`, but I'm wondering how that `handle.await` does not block the converter until the spawned task is complete.